### DISCUSSION TITLE
Column chart with single data not showing label

### DISCUSSION THREAD

**Discussion by Didi211** [patrykandpatrick/vico#543] [Questions]
In this code 
```
val data = readings
            .map { it.timeMeasured to it.reading.toFloat() } // x and y values for graph
            .associate { (date, yValue) -> date to yValue }

        // converting data to float value
        val xValuesToDates = data.keys.associateBy { it.truncatedTo(ChronoUnit.HOURS).toEpochSecond(ZoneOffset.UTC).toFloat() }
        // creating entries - x and y's for graph
        val chartEntryModel = entryModelOf(xValuesToDates.keys.zip(data.values, ::entryOf))

        val count = chartEntryModel.entries[0].count() // will define if is column chart or line
        val chartType = if (count == 1) columnChart() else lineChart()
        val valueFormatter = when (graphPeriods) {
            GraphPeriods.Hours -> AxisHourFormatter()
            GraphPeriods.Months -> AxisMonthFormatter()
            GraphPeriods.Years -> AxisYearFormatter()
        }
        Chart(
            modifier = Modifier
                .clip(RoundedCornerShape(10.dp))
                .background(Color.White)
                .padding(5.dp),
            model = chartEntryModel,
            chart = chartType,
            startAxis = rememberStartAxis(),
            bottomAxis = rememberBottomAxis(
                valueFormatter = valueFormatter
            ),
        )
```
I am trying to show only column if that is the case. But label is not showing in that situation. Has anyone encountered this problem? I tried to look up in docs but couldn't find anything. Formatter classes are doing their job properly and are returning good formated value.

**Comment by Gowsky** [comment]
Hi @Didi211, could you provide a minimal reproducible example and possibly a screenshot?

**Reply by Didi211** [reply]
If you by that mean a whole composable that I am trying to draw, then here it is

```
@Composable
fun GraphChart(readings: List<GraphReading> = emptyList(), graphPeriods: GraphPeriods = GraphPeriods.Hours,) {
    if (readings.isEmpty()) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) {
            Text(text = "Data will be available after a while." )
        }
        return
    }
    ProvideChartStyle(
        chartStyle = ChartStyle.fromColors(
            axisLabelColor = FontColor,
            axisGuidelineColor = FontColor,
            axisLineColor = FontColor,
            entityColors = listOf(Secondary),
            elevationOverlayColor = BackgroundVariant
        ),
    ) {
        val data = readings
            .map { it.timeMeasured to it.reading.toFloat() } // x and y values for graph
            .associate { (date, yValue) -> date to yValue }

        // converting data to float value
        val xValuesToDates = data.keys.associateBy { it.truncatedTo(ChronoUnit.HOURS).toEpochSecond(ZoneOffset.UTC).toFloat() }
        // creating entries - x and y's for graph
        val chartEntryModel = entryModelOf(xValuesToDates.keys.zip(data.values, ::entryOf))


        val valueFormatter = when (graphPeriods) {
            GraphPeriods.Hours -> AxisHourFormatter()
            GraphPeriods.Months -> AxisMonthFormatter()
            GraphPeriods.Years -> AxisYearFormatter()
        }
        --val count = chartEntryModel.entries[0].count() // will define if is column chart or line
        //val chartType = if (count == 1) columnChart() else lineChart()
        Chart(
            modifier = Modifier
                .clip(RoundedCornerShape(10.dp))
                .background(Color.White)
                .padding(5.dp),
            model = chartEntryModel,
            chart = columnChart(),
            startAxis = rememberStartAxis(),
            bottomAxis = rememberBottomAxis(
                valueFormatter = valueFormatter,
            ),
        )
    }
}
```

![image](https://github.com/patrykandpatrick/vico/assets/82868612/bede4a36-ea6c-4706-ae87-afae322095fb)
![image](https://github.com/patrykandpatrick/vico/assets/82868612/84c8a8f6-6eed-49dd-9d7c-d75985448c3e)
On the second picture **readings** variable only has one element

**Reply by Gowsky** [reply]
MRE must be self-contained, so it must contain the data you're using to display in the chart. In addition, it must be minimal, thus it should use only code relevant to the bug.
Also, what Vico version are you using?

**Reply by Didi211** [reply]
I am using the latest version available on wiki

**Reply by patrickmichalik** [reply]
In our testing, @Didi211, no such issue occurs, as can be seen below. We’ll thus need a self-contained, runnable example, as mentioned by @Gowsky.

<img src="https://github.com/patrykandpatrick/vico/assets/120058021/0cca8288-8599-46f7-a122-ddc1c57b217d" width="720" />

**Reply by SSHshadow222** [reply]
I was able to reproduce the issue on versions `2.0.0-alpha.14` and `2.0.0-alpha.16` by setting the only series x value to a higher value, like `8_388_610`. Here's a minimal code example:

```kotlin
@Composable
private fun OneColumnChart() {
    val modelProducer = remember { CartesianChartModelProducer.build() }

    CartesianChartHost(
        chart = rememberCartesianChart(
            rememberColumnCartesianLayer(),
            startAxis = rememberStartAxis(),
            bottomAxis = rememberBottomAxis(),
            topAxis = rememberTopAxis(),
            endAxis = rememberEndAxis()
        ),
        modelProducer = modelProducer
    )

    LaunchedEffect(Unit) {
        withContext(Dispatchers.Default) {
            modelProducer.tryRunTransaction {
                columnSeries {
                    series(
                        x = listOf(8_388_610),
                        y = listOf(1)
                    )
                }
            }
        }
    }
}
```

Here's the result:

![Screenshot_20240420-223044 (1)](https://github.com/patrykandpatrick/vico/assets/113729758/d1c63a54-0044-4f89-840c-fc3d35271884)

If I change the value to `8_388_609` or `8_388_611`, it works as expected:

![Screenshot_20240420-223009](https://github.com/patrykandpatrick/vico/assets/113729758/010f9f59-fbe2-42f3-977b-96df38c3fcd1)

---

![Screenshot_20240420-230854](https://github.com/patrykandpatrick/vico/assets/113729758/af0673b2-27b6-4212-b951-5a5a1226c32b)

I don't know why it doesn't work for certain numbers. I picked those by experimenting. However, I initially reached this issue by using milliseconds from an epoch as my x values, and as you can imagine, the values were even higher. So, I would assume that this has to do with big numbers.

**Reply by mustfaunlu** [reply]
When you have a single value in the chart, the column is too big. How do I reduce the width of the column?

**Reply by Gowsky** [reply]
Hi @mustfaunlu, you can change the zoom settings to achieve the desired result. This is covered [in the wiki](https://patrykandpatrick.com/vico/wiki/cartesian-charts/hosts#scroll-and-zoom). For example, you can set the minimum and initial zoom factors to 1. With Vico 2 and Jetpack Compose, you’d do this as follows:
```kotlin
CartesianChartHost(
    ...
    zoomState = rememberVicoZoomState(initialZoom = Zoom.static(1f), minZoom = Zoom.static(1f)),
)
```


Next time, please create a dedicated discussion, as your question is off-topic.

**Reply by Gowsky** [reply]
Hi @SSHshadow222, I deleted my previous comment, as it wasn't entirely correct. This bug is a result of `Float` imprecision. We’ll be switching to `Double` in the future and it should help the case. Please see #623 for more.

In general, such large _x_ values are unnecessary. We recommend using smaller _x_ values (for example, by shifting everything so that the first value is zero) and mapping them as required in `CartesianValueFormatter`s (with the use of `ExtraStore`). The solution described [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/axes#dates) uses days from the Epoch. With milliseconds from the Epoch, you end up with values that are too large. You can address this by subtracting the first timestamp from all of the timestamps, saving it as an extra, and adding it back during formatting. Also, if you only need steps of five milliseconds, for example, then you can divide all of the timestamps by five.

If you need any help, please let us know what your use case is so we can suggest a specific approach.
